core/trino-main/src/main/java/io/trino/execution/querystats/QueryRuleStats.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.querystats;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.stats.TimeDistribution;
+import io.airlift.stats.TimeDistribution.TimeDistributionSnapshot;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRuleStats
+{
+    private final String rule;
+    private long invocations;
+    private long applications;
+    private final TimeDistribution time = new TimeDistribution(TimeUnit.MICROSECONDS);
+    private long totalTime;
+    private long failures;
+
+    public QueryRuleStats(String rule)
+    {
+        this.rule = rule;
+    }
+
+    public void record(long nanos, boolean match)
+    {
+        if (match) {
+            applications += 1;
+        }
+
+        invocations += 1;
+        totalTime += nanos;
+        time.add(nanos);
+    }
+
+    public void recordFailure()
+    {
+        failures += 1;
+    }
+
+    @JsonProperty
+    public String getRule()
+    {
+        return rule;
+    }
+
+    @JsonProperty
+    public long getInvocations()
+    {
+        return invocations;
+    }
+
+    public TimeDistributionSnapshot getTime()
+    {
+        return time.snapshot();
+    }
+
+    public long getApplications()
+    {
+        return applications;
+    }
+
+    public long getFailures()
+    {
+        return failures;
+    }
+
+    public long getTotalTime()
+    {
+        return totalTime;
+    }
+
+    public QueryRuleStatsSnapshot snapshot()
+    {
+        return new QueryRuleStatsSnapshot(rule, invocations, applications, time.snapshot(), totalTime, failures);
+    }
+
+    public static class QueryRuleStatsSnapshot
+    {
+        private final String rule;
+        private final long invocations;
+        private final long applications;
+        private final TimeDistributionSnapshot time;
+        private final long totalTime;
+        private final long failures;
+
+        @JsonCreator
+        public QueryRuleStatsSnapshot(
+                @JsonProperty("rule") String rule,
+                @JsonProperty("invocations") long invocations,
+                @JsonProperty("applications") long applications,
+                @JsonProperty("time") TimeDistributionSnapshot time,
+                @JsonProperty("totalTime") long totalTime,
+                @JsonProperty("failures") long failures)
+        {
+            this.rule = requireNonNull(rule, "rule is null");
+            checkArgument(invocations >= 0, "invocations is negative");
```

Comment:
nit: I don't think it can happen, since it's internal class, we could maybe skip it